python/mlx/nn/utils.py

```diff
@@ -31,3 +34,53 @@ def wrapped_value_grad_fn(*args, **kwargs):
         return value, grad
 
     return wrapped_value_grad_fn
+
+
+def maybe_checkpoint(fn: Callable):
```

Comment:
What's the usage with `maybe_checkpoint`? Why not always checkpoint the module when you use this wrapper?

And if you don't want a check pointed module, don't call it?

Then you don't really need to pass in kwargs or add an attribute to the module which I think adds a lot keep track of in the API.

Reply:
Concretely, usage like the following seems a bit more intuitive to me.

```
m = nn.chekcpoint(m) if args.checkpoint else m
outputs = m(inputs)
```

And I think we could make it so that it only works with modules to simplify a bit?

Reply:
So I agree that we should add a `nn.checkpoint` that does what `mx.checkpoint` does (what it will do) but for modules.

Regarding `maybe_checkpoint`, I think it is quite common to want to optionally enable checkpointing. For instance the same code may be running models from 100M to 10B parameters and for the 10B you almost certainly need checkpointing but for the 100M you almost certainly don't. It is like having a `Checkpointable` mixin module.

Reply:
I would do it the way it is in the transformer now. The module is checkpointable and its container passes checkpoint to it or we can loop and activate checkpointing with something like

```python
def set_checkpointing(m):
    if hasattr(m, "checkpoint"):
        m.checkpoint = True

model.apply_to_modules(set_checkpointing)
```

For the LoRA example since we are looping to set LoRA layers I simply added a loop that did the above, something like

```python
for l in model.layers:
    l.checkpoint = True
```

Reply:
Why not just do something like:

```
if config.checkpoint:
  model.layers = [nn.checkpoint(l) for l in model.layers]
```

You could do the same with `model.apply_to_modules` right?

Is there something that you get with the attribute that is hard to do otherwise?

Reply:
> I would do it the way it is in the transformer now.

It seems like it would work just as well in that example to checkpoint in the constructor based on the flag and wouldn't require the addition of `maybe_checkpoint` and the `checkpoint` attribute...but I might be missing something in terms of `nn.checkpoint` works.

Reply:
Well ... removing the checkpointing. I know it is kind of unlikely. I am not married to the attribute... I think it might be worth doing something else or even just removing the attribute and relying on the kwarg.